packages/block-editor/src/store/private-selectors.js

```diff
@@ -581,3 +582,52 @@ export function isZoomOutMode( state ) {
 export function getSectionRootClientId( state ) {
 	return state.settings?.[ sectionRootClientIdKey ];
 }
+
+/**
+ * Retrieves the editing mode of a pattern block and its children.
+ *
+ * @param {Object} state    Global application state.
+ * @param {string} clientId The block client ID.
+ *
+ * @return {string} The block editing mode.
+ */
+export function getPatternBlockEditingMode( state, clientId ) {
+	const parentPatternCount = getParentPatternCount( state, clientId );
+
+	// If there are no parent patterns, use the block's own editing mode.
+	if ( parentPatternCount === 0 ) {
+		return state.blockEditingModes.get( clientId );
+	}
+
+	// Disable nested patterns.
+	if ( parentPatternCount > 1 ) {
+		return 'disabled';
+	}
+
+	// If the block has a binding of any kind, allow content only editing.
+	const attributes = getBlockAttributes( state, clientId );
+	if ( Object.keys( attributes.metadata?.bindings ?? {} )?.length > 0 ) {
+		return 'contentOnly';
+	}
+
+	// Otherwise, the block is part of the pattern source and should not be editable.
+	return 'disabled';
+}
+
+/**
+ * Retrieves the number of parent pattern blocks.
+ *
+ * @param {Object} state    Global application state.
+ * @param {string} clientId The block client ID.
+ *
+ * @return {number} The number of parent pattern blocks.
+ */
+export function getParentPatternCount( state, clientId ) {
```

Comment:
Might be an idea to make this one memoized, and also check that the memoization of `getBlockEditingMode` is still correct 🤔